boa_engine/src/syntax/parser/expression/primary/mod.rs

```diff
@@ -243,3 +256,304 @@ where
+                TokenKind::Punctuator(Punctuator::CloseParen) => {
+                    let span = next.span();
+                    cursor.next(interner).expect("token disappeared");
+                    break span;
+                }
+                TokenKind::Punctuator(Punctuator::Comma) => {
+                    let span = next.span();
+                    cursor.next(interner).expect("token disappeared");
+                    if let Some(token) = cursor.next_if(Punctuator::CloseParen, interner)? {
+                        tailing_comma = Some(span);
+                        break token.span();
+                    }
+                }
+                TokenKind::Punctuator(Punctuator::Spread) => {
+                    cursor.next(interner).expect("token disappeared");
+                    let next = cursor.peek(0, interner)?.ok_or(ParseError::AbruptEnd)?;
+                    match next.kind() {
+                        TokenKind::Punctuator(Punctuator::OpenBlock) => {
+                            let bindings =
+                                ObjectBindingPattern::new(self.allow_yield, self.allow_await)
+                                    .parse(cursor, interner)?;
+                            expressions.push(InnerExpression::SpreadObject(bindings));
+                        }
+                        TokenKind::Punctuator(Punctuator::OpenBracket) => {
+                            let bindings =
+                                ArrayBindingPattern::new(self.allow_yield, self.allow_await)
+                                    .parse(cursor, interner)?;
+                            expressions.push(InnerExpression::SpreadArray(bindings));
+                        }
+                        _ => {
+                            let binding =
+                                BindingIdentifier::new(self.allow_yield, self.allow_await)
+                                    .parse(cursor, interner)?;
+                            expressions.push(InnerExpression::SpreadBinding(binding));
+                        }
+                    }
+                }
+                _ => {
+                    let expression =
+                        Expression::new(self.name, true, self.allow_yield, self.allow_await)
+                            .parse(cursor, interner)?;
+                    expressions.push(InnerExpression::Expression(expression));
+                }
+            }
+        };
+
+        let is_arrow = if let Some(TokenKind::Punctuator(Punctuator::Arrow)) =
+            cursor.peek(0, interner)?.map(Token::kind)
+        {
+            !cursor.peek_expect_is_line_terminator(0, interner)?
+        } else {
+            false
+        };
+
+        // If the next token is not an arrow, we know that we must parse a parenthesized expression.
+        if !is_arrow {
+            if let Some(span) = tailing_comma {
+                return Err(ParseError::unexpected(
+                    Punctuator::Comma.to_string(),
```

Comment:
Is the `to_string()` needed?

Reply:
It was, because we take `Into<Box<str>>` in `ParseError::unexpected`. I implemented `From<Punctuator> for Box<str>` and now it works without.

Reply:
Thanks! I didn't know that!